### PR TITLE
[torch][ao] Add customizable loss function to NodeAccuracySummary

### PR DESCRIPTION
**Pull request by dulinriley**
Summary:
Add a customizable loss function callback to NodeAccuracySummary to
allow users to pass in their own loss function.

Also, fix some type errors and propagate better exception messages when
unexpected tensor comparisons occur. Finally, enhance the robustness of
`generate_numeric_debug_handle` in the case where it is called multiple
times on the same model, by avoiding reuse of the same IDs.

Test Plan: Added a test for this case in `test_numeric_debugger`.

Reviewed By: jerryzh168

Differential Revision: D62898297

### CODE REVIEW

**Review by jerryzh168 — APPROVED** [review]

**Comment by dulinriley** [comment]
@huydhn do you have a link to show what was broken that caused this to get reverted?

**Comment by dulinriley** [comment]
I think I fixed the assert by just allowing list or tuple.
@pytorchbot merge

**Comment by dulinriley** [comment]
Looks like this landed now with fixed tests. Let me know if it causes any issues
